### DISCUSSION TITLE
How to mock wagmi for testing purposes?

### DISCUSSION THREAD

**Discussion by stefan-girlich** [wevm/wagmi#577] [Q&A]
I am using wagmi (mostly `@wagmi/core`) in custom hooks within my React application. Now I want to mock wallet/account data managed by wagmi so I can unit-test the logic in these custom hooks.

The wagmi codebase has some boilerplate code (see [here](https://github.com/tmm/wagmi/tree/main/packages/core/test) ), but does not seem to contain tests.

Does anyone have experience with mocking wagmi that they could share? I think it would also be beneficial to include testing/mocking instructions in the documentation.

**Comment by tmm** [comment]
There are a bunch of tests colocated with the source files! For example, `@wagmi/core`'s [`connect` action](https://github.com/tmm/wagmi/blob/main/packages/core/src/actions/accounts/connect.test.ts).

We run integration tests (with hardhat) more than unit tests with mocking so we can verify state changes. If you want to just test accounts (without reading/writing on-chain data), you can use the `MockConnector` and pass that to your wagmi `Client`. Adding testing to the documentation is on the list, but not as high priority as other items.

**Reply by stefan-girlich** [reply]
Thanks for your quick answer, @tmm 

For my particular use case (integration testing via Jest), I am trying to extract my logic to a pure function in which I want to call `signTypedData` and `writeContract` against Rinkeby.
Could you recommend a way to create a static signer (e.g. set their key pair from environment variables) without going through Metamask?  [Your test code](https://github.com/tmm/wagmi/blob/e6092850aea72fb5b2985e6333190dc0169f5d27/packages/core/test/utils.ts#L151) looks promising; but maybe there is a simpler option

**Reply by stefan-girlich** [reply]
Created a more detailed discussion regarding invoking contract functions in integration tests: https://github.com/tmm/wagmi/discussions/579

**Comment by pavel-lens** [comment]
@tmm I'm looking at the `MockConnector` after installing the wagmi package, but I don't think that's exported. I just upgraded to latest released version. Did you mean that we should copy the `MockConnector` from your implementation or did you not realize that it's not exported? Thanks!

**Reply by pavel-lens** [reply]
Ah, I found it in `@wagmi/core` package.

```ts
import { MockConnector } from '@wagmi/core/connectors/mock';
```

**Reply by tmm** [reply]
Will export it from `wagmi` too!

**Reply by Na3aga** [reply]
Hello, I also have a problem with writing e2e tests with cypress for custom hooks, that use wagmi. 
@svitekpavel maybe you can share your experience with me?

**Reply by pavel-lens** [reply]
The closest I've come to E2E testing is with Synpress which is a wrapper of Cypress with Metamask, but I didn't find it very stable either. Even if you look at the example tests from different projects which they mentioned in the documentation, there's only handful of those tests which indicates that running the tests either doesn't work or is tedious or both. I had problems with Chrome not starting every time, the Metamask installation process failing for unknown reasons and series of timeouts in multiple places.

https://github.com/Synthetixio/synpress

**Reply by pavel-lens** [reply]
@Na3aga have a look here, the maintainer created a sample repo with both a) component tests and b) E2E test with Playwright
Rg. React component testing with wagmi lib + E2E testing, I just god reply from the lib maintainer, gonna send him $$ tip for this .. consider sending a tip as well if you are using wagmi and this helps you build better tests 🙏🤲
https://github.com/wagmi-dev/wagmi/discussions/774#discussioncomment-3390453
